spec/core/ics-003-connection-semantics/UPGRADES.md

```diff
@@ -0,0 +1,407 @@
+
+As discussed, during the upgrade handshake a chain may cancel the upgrade by writing `UPGRADE_ERR` into the `UpgradeStatus` and restoring the original connection to `OPEN`. The counterparty must then restore its connection to `OPEN` as well. A relayer can facilitate this by calling `CancelConnectionUpgrade`
+
+```typescript
+function cancelConnectionUpgrade(
+    identifier: Identifer,
+    counterpartyUpgradeStatus: UpgradeStatus,
+    proofUpgradeStatus: CommitmentProof,
+    proofHeight: Height,
+) {
+    // current connection is in UPGRADE_INIT or UPGRADE_TRY
+    currentConnection = provableStore.get(connectionPath(identifier))
+    abortTransactionUnless(connection.state == (UPGRADE_INIT || UPGRADE_TRY))
+
+    abortTransactionUnless(counterpartyUpgradeStatus.state == UPGRADE_ERR)
+
+    abortTransactionUnless(currentConnection.client.verifyUpgradeStatus(proofHeight, proofUpgradeStatus, counterpartyUpgradeStatus))
+
+    // cancel upgrade
+    // and restore original conneciton
+    // delete unnecessary state
+    originalConnection = privateStore.get(restorePath(identifier))
+    provableStore.set(connectionPath(identifier), originalConnection)
+    provableStore.delete(statusPath(identifier))
+    privateStore.delete(restorePath(identifier))
+}
+```
+
+### Timeout Upgrade Process
+
+It is possible for the connection upgrade process to stall indefinitely on UPGRADE_TRY if the UPGRADE_TRY transaction simply cannot pass on the counterparty; for example, the upgrade feature may not be enabled on the counterparty chain.
+
+In this case, we do not want the initializing chain to be stuck indefinitely in the `UPGRADE_INIT` step. Thus, the `UpgradeInit` message will contain a `TimeoutHeight` and `TimeoutTimestamp`. The counterparty chain is expected to reject `UpgradeTry` message if the specified timeout has already elapsed.
+
+A relayer must then submit an `UpgradeTimeout` message to the initializing chain which proves that the counterparty is still in its original state. If the proof succeeds, then the initializing chain shall also restore its original connection and cancel the upgrade.
+
+```typescript
+function timeoutConnectionUpgrade(
+    identifier: Identifier,
+    counterpartyConnection: ConnectionEnd,
+    proofConnection: CommitmentProof,
+    proofHeight: Height,
+) {
+    upgradeStatus = provableStore.get(statusPath(identifier))
+
+    // proof must be from a height after timeout has elapsed. Either timeoutHeight or timeoutTimestamp must be defined.
+    // if timeoutHeight is defined and proof is from before timeout height
+    // then abort transaction
+    abortTransactionUnless(upgradeStatus.TimeoutHeight.IsZero() || proofHeight >= upgradeStatus.TimeoutHeight)
+    // if timeoutTimestamp is defined then the consensus time from proof height must be greater than timeout timestamp
+    connection = getConnection(identifer)
+    consensusState = getConsensusState(connection.clientIdentifer, proofHeight)
```

Comment:
More of a general comment, `getConsensusState(id, height)` is not defined in the IBC specs.

Reply:
queryConsensusState exists, so i replaced it with that